### PR TITLE
pd: switch to grpc-rs

### PR DESCRIPTION
**Pull request by overvenus**
This replace [grpc-rust](https://github.com/stepancheg/grpc-rust/) with [grpc-rs](https://github.com/pingcap/grpc-rs) (Written on top of shared C core library [grpc/src/core](https://github.com/grpc/grpc)).

Pending

 - [x] ~~pingcap/grpc-rs#5~~
 - [x] pingcap/grpc-rs#13
 - [ ] pingcap/kvproto#157

### CODE REVIEW

File: src/pd/async/client.rs [hunk 1]
```diff
@@ -39,11 +42,12 @@ impl RpcClient {
             })
             .collect();
 
-        let (client, members) = try!(validate_endpoints(&endpoints));
+        let env = Arc::new(Environment::new(1));
```
Comment (siddontang) [inline src/pd/async/client.rs]: if we use different Env for PD and server, we should pass different name for the Env loop thread. 

So it is better to pass name when creating Environment. @BusyJay

**Comment by siddontang** [comment]
PTAL @BusyJay @disksing

File: src/pd/async/util.rs [hunk 2]
```diff
@@ -274,37 +281,32 @@ pub fn validate_endpoints(endpoints: &[String]) -> Result<(PDAsyncClient, GetMem
...
-        .and_then(|client| {
-            // try request.
-            match client.GetMembers(GetMembersRequest::new()).wait() {
-                Ok(resp) => Ok((client, resp)),
-                Err(e) => Err(e),
-            }
-        })
-        .map_err(Error::Grpc)
+    let addr = if addr.starts_with("http://") {
+        &addr[7..]
```
Comment (ngaut) [inline src/pd/async/util.rs]: Define 7 as constant.

Comment (disksing) [inline src/pd/async/util.rs]: Try this: https://doc.rust-lang.org/std/string/struct.String.html#method.trim_left_matches

File: src/pd/async/util.rs [hunk 3]
```diff
@@ -24,35 +24,41 @@ use futures::future::{loop_fn, Loop};
...
-    client: PDAsyncClient,
+    env: Arc<Environment>,
+    client: PDClient,
     members: GetMembersResponse,
 
     last_update: Instant,
 }
 
+unsafe impl Send for Inner {}
+unsafe impl Sync for Inner {}
```
Comment (disksing) [inline src/pd/async/util.rs]: Is it safe to impl Sync? cc @BusyJay

File: Cargo.toml [hunk 4]
```diff
@@ -83,7 +84,8 @@ default-features = false
 features = ["nightly", "push"]
 
 [dependencies.grpc]
-git = "https://github.com/stepancheg/grpc-rust.git"
+git = "https://github.com/pingcap/grpc-rs.git"
+branch = "ov/compiler-dev"
```
Comment (siddontang) [inline Cargo.toml]: Please use busyjay/dev

**Comment by siddontang** [comment]
@disksing 

Seem we can use base:grpc branch here now.

File: Cargo.toml [hunk 5]
```diff
@@ -83,7 +84,8 @@ default-features = false
 features = ["nightly", "push"]
 
 [dependencies.grpc]
-git = "https://github.com/stepancheg/grpc-rust.git"
+git = "https://github.com/pingcap/grpc-rs.git"
+branch = "busyjay/dev"
```
Comment (siddontang) [inline Cargo.toml]: use static link here, so we don't need to install c grpc manually in CI.

File: src/pd/async/client.rs (hunk 1 above)
Comment (siddontang) [inline src/pd/async/client.rs]: please use the Env builder with a customized thread name.

Comment (siddontang) [inline src/pd/async/client.rs]: define a const var for 1

File: src/raftstore/store/worker/apply.rs [hunk 6]
```diff
@@ -1293,6 +1293,9 @@ impl Runner {
         let region_id = s.region.get_id();
         let term = s.term;
         let delegate = ApplyDelegate::from_registration(self.db.clone(), s);
+        info!("{} register to apply delegates at term {}",
```
Comment (siddontang) [inline src/raftstore/store/worker/apply.rs]: seem we should merge master to the grpc branch first.

File: src/pd/errors.rs [hunk 7]
```diff
@@ -31,7 +31,7 @@ quick_error!{
             description("cluster not bootstrap error")
             display("cluster {} is not bootstrapped", cluster_id)
         }
-        Grpc(err: ::grpc::error::GrpcError) {
+        Grpc(err: ::grpc::Error) {
```
Comment (siddontang) [inline src/pd/errors.rs]: use grpc::Error as GrpcError

File: src/pd/async/util.rs [hunk 8]
```diff
@@ -205,13 +208,12 @@ impl<Req, Resp, F> Request<Req, Resp, F>
...
 pub fn sync_request<F, R>(client: &LeaderClient, retry: usize, func: F) -> Result<R>
-    where F: Fn(&PDAsyncClient) -> GrpcFutureSend<R>
+    where F: Fn(&PDClient) -> GrpcResult<R>
 {
     for _ in 0..retry {
         let r = {
             let _timer = PD_SEND_MSG_HISTOGRAM.start_timer(); // observe on dropping.
-            let f = func(&client.inner.rl().client);
-            f.wait()
+            func(&client.inner.rl().client).map_err(Error::Grpc)
```
Comment (siddontang) [inline src/pd/async/util.rs]: no wait now?

File: src/pd/errors.rs (hunk 7 above)
Comment (BusyJay) [inline src/pd/errors.rs]: I think it's ok to namespace error.

File: src/pd/async/util.rs (hunk 8 above)
Comment (overvenus) [inline src/pd/async/util.rs]: It not a future any more, `grpc::Result<_>`.

Comment (siddontang) [inline src/pd/async/util.rs]: got it

**Comment by siddontang** [comment]
PTAL @BusyJay @disksing

**Comment by disksing** [comment]
LGTM.

**Pull request merged**